File: fixtures/proc-macro/tests/bindings/test_proc_macro.swift
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import Foundation
import proc_macro

let one = makeOne(inner: 123)
assert(one.inner == 123)
assert(oneInnerByRef(one: one) == 123)
assert(one.getInnerValue() == 123)

let two = Two(a: "a")
assert(takeTwo(two: two) == "a")

let rwb = RecordWithBytes(someBytes: Data([1, 2, 3]))
assert(takeRecordWithBytes(rwb: rwb) == Data([1, 2, 3]))

var obj = Object()
obj = Object.namedCtor(arg: 1)
assert(obj.isHeavy() == .uncertain)
let obj2 = Object()
assert(obj.isOtherHeavy(other: obj2) == .uncertain)

let traitImpl = obj.getTrait(inc: nil)
assert(traitImpl.concatStrings(a: "foo", b: "bar") == "foobar")
assert(obj.getTrait(inc: traitImpl).concatStrings(a: "foo", b: "bar") == "foobar")
assert(concatStringsByRef(t: traitImpl, a: "foo", b: "bar") == "foobar")

let traitImpl2 = obj.getTraitWithForeign(inc: nil)
assert(traitImpl2.name() == "RustTraitImpl")
assert(obj.getTraitWithForeign(inc: traitImpl2).name() == "RustTraitImpl")

assert(enumIdentity(value: .true) == .true)

// just make sure this works / doesn't crash
let three = Three(obj: obj)

assert(makeZero().inner == "ZERO")
assert(makeRecordWithBytes().someBytes == Data([0, 1, 2, 3, 4]))
assert(join(parts: ["a", "b", "c"], sep: ":") == "a:b:c")

do {
    try alwaysFails()
    fatalError("alwaysFails should have thrown")
} catch BasicError.OsError {
}

try! obj.doStuff(times: 5)

do {
    try obj.doStuff(times: 0)
    fatalError("doStuff should throw if its argument is 0")
} catch FlatError.InvalidInput {
}

struct SomeOtherError: Error { }

// Defaults

let recordWithDefaults = RecordWithDefaults(noDefaultString: "Test")
assert(recordWithDefaults.noDefaultString == "Test")
assert(recordWithDefaults.boolean == true)
assert(recordWithDefaults.integer == 42)
assert(recordWithDefaults.floatVar == 4.2)
assert(recordWithDefaults.vec == [])
assert(recordWithDefaults.optVec == nil)
assert(recordWithDefaults.optInteger == 42)
assert(recordWithDefaults.customInteger == 42)

// Implicit defaults
let recordWithImplicitDefaults = RecordWithImplicitDefaults()
assert(recordWithImplicitDefaults.boolean == false)
assert(recordWithImplicitDefaults.int8 == 0)
assert(recordWithImplicitDefaults.uint8 == 0)
assert(recordWithImplicitDefaults.int16 == 0)
assert(recordWithImplicitDefaults.uint16 == 0)
assert(recordWithImplicitDefaults.int32 == 0)
assert(recordWithImplicitDefaults.uint32 == 0)
assert(recordWithImplicitDefaults.int64 == 0)
assert(recordWithImplicitDefaults.uint64 == 0)
assert(recordWithImplicitDefaults.afloat == 0.0)
assert(recordWithImplicitDefaults.adouble == 0.0)
assert(recordWithImplicitDefaults.vec == [])
assert(recordWithImplicitDefaults.map == [:])
assert(recordWithImplicitDefaults.someBytes == Data([]))
assert(recordWithImplicitDefaults.optInt32 == nil)
assert(recordWithImplicitDefaults.customInteger == 0)

// defaults in function args
assert(doubleWithDefault() == 42)
assert(sumWithDefault(num1: 1) == 1)
assert(sumWithDefault(num1: 1, num2: 2) == 3)

let objWithDefaults = ObjectWithDefaults()
assert(objWithDefaults.addToNum() == 42)
assert(objWithDefaults.addToImplicitNum() == 30)
assert(objWithDefaults.addToImplicitNum(other: 1) == 31)

// Traits

final class SwiftTestCallbackInterface : TestCallbackInterface {
    func doNothing() { }

    func add(a: UInt32, b: UInt32) -> UInt32 {
        return a + b;
    }

    func `optional`(a: Optional<UInt32>) -> UInt32 {
        return a ?? 0;
    }

    func withBytes(rwb: RecordWithBytes) -> Data {
        return rwb.someBytes
    }

    func tryParseInt(value: String) throws -> UInt32 {
        if (value == "force-unexpected-error") {
            // raise an error that's not expected
            throw SomeOtherError()
        }
        let parsed = UInt32(value)
        if parsed != nil {
            return parsed!
        } else {
            throw BasicError.InvalidInput
        }
    }

    func callbackHandler(h: Object) -> UInt32 {
        return h.takeError(e: BasicError.InvalidInput)
    }

    func getOtherCallbackInterface() -> OtherCallbackInterface {
        SwiftTestCallbackInterface2()
    }
}

final class SwiftTestCallbackInterface2 : OtherCallbackInterface {
    func multiply(a: UInt32, b: UInt32) -> UInt32 {
        return a * b;
    }
}

callCallbackInterface(cb: SwiftTestCallbackInterface())

assert(getMixedEnum(v: nil) == .int(1))
assert(getMixedEnum(v: MixedEnum.none) == .none)
assert(getMixedEnum(v: MixedEnum.string("hello")) == .string("hello"))
switch MixedEnum.string("hello") {
    case let .string(s):
        assert(s == "hello")
    default:
        assert(false)
}

switch MixedEnum.both("hello", 1) {
    case let .both(s, i):
        assert(s == "hello")
        assert(i == 1)
    default:
        assert(false)
}

switch MixedEnum.all(s: "string", i: 2) {
    case let .all(s, i):
        assert(s == "string")
        assert(i == 2)
    default:
        assert(false)
}

assert(getMixedEnum(v: MixedEnum.vec(["hello"])) == .vec(["hello"]))

// check autogenerated CaseIterable
assert(MaybeBool.allCases.map({ "\($0)" }).joined(separator: ", ") == "true, false, uncertain")
assert(ReprU8.allCases.map({ "\($0)" }).joined(separator: ", ") == "one, three")
assert(SimpleError.allCases.map({ "\($0)" }).joined(separator: ", ") == "A, B")

File: fixtures/proc-macro/tests/bindings/test_proc_macro.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

from proc_macro import *
from proc_macro import TraitWithForeignImpl

one = make_one(123)
assert one.inner == 123
assert one_inner_by_ref(one) == 123
assert one.get_inner_value() == 123

two = Two(a="a")
assert take_two(two) == "a"

rwb = RecordWithBytes(some_bytes=bytes([1,2,3]))
assert take_record_with_bytes(rwb) == bytes([1,2,3])

obj = Object()
obj = Object.named_ctor(1)
assert obj.is_heavy() == MaybeBool.UNCERTAIN
obj2 = Object()
assert obj.is_other_heavy(obj2) == MaybeBool.UNCERTAIN

robj = Renamed()
assert(robj.func())
assert(rename_test())

trait_impl = obj.get_trait(None)
assert isinstance(trait_impl, Trait)
assert trait_impl.concat_strings("foo", "bar") == "foobar"
assert obj.get_trait(trait_impl).concat_strings("foo", "bar") == "foobar"
assert concat_strings_by_ref(trait_impl, "foo", "bar") == "foobar"
assert StructWithTrait("me").concat_strings("foo", "bar") == "me: foobar"

trait_impl2 = obj.get_trait_with_foreign(None)
# This is an instance of `TraitWithForeignImpl` - `TraitWithForeign` is used primarily for subclassing.
assert isinstance(trait_impl2, TraitWithForeignImpl)
assert trait_impl2.name() == "RustTraitImpl"
assert obj.get_trait_with_foreign(trait_impl2).name() == "RustTraitImpl"

assert enum_identity(MaybeBool.TRUE) == MaybeBool.TRUE

# just make sure this works / doesn't crash
three = Three(obj=obj)

assert(make_zero().inner == "ZERO")
assert(make_record_with_bytes().some_bytes == bytes([0, 1, 2, 3, 4]))

assert(make_hashmap(1, 2) == {1: 2})
d = {1: 2}
assert(return_hashmap(d) == d)

assert(join(["a", "b", "c"], ":") == "a:b:c")

try:
    always_fails()
except BasicError.OsError:
    pass
else:
    raise Exception("always_fails should have thrown")

obj.do_stuff(5)

try:
    obj.do_stuff(0)
except FlatError.InvalidInput:
    pass
else:
    raise Exception("do_stuff should throw if its argument is 0")


# Defaults

record_with_defaults = RecordWithDefaults(no_default_string="Test")
assert(record_with_defaults.no_default_string == "Test")
assert(record_with_defaults.boolean == True)
assert(record_with_defaults.integer == 42)
assert(record_with_defaults.float_var == 4.2)
assert(record_with_defaults.vec == [])
assert(record_with_defaults.opt_vec == None)
assert(record_with_defaults.opt_integer == 42)
assert(record_with_defaults.custom_integer == 42)

assert(RecordWithDefaults(no_default_string="", vec=["oops"]) == RecordWithDefaults(no_default_string="", vec=["oops"]))

record_with_implicit_defaults = RecordWithImplicitDefaults()
assert(record_with_implicit_defaults.boolean == False)
assert(record_with_implicit_defaults.int8 == 0)
assert(record_with_implicit_defaults.uint8 == 0)
assert(record_with_implicit_defaults.int16 == 0)
assert(record_with_implicit_defaults.uint16 == 0)
assert(record_with_implicit_defaults.int32 == 0)
assert(record_with_implicit_defaults.uint32 == 0)
assert(record_with_implicit_defaults.int64 == 0)
assert(record_with_implicit_defaults.uint64 == 0)
assert(record_with_implicit_defaults.afloat == 0.0)
assert(record_with_implicit_defaults.adouble == 0.0)
assert(record_with_implicit_defaults.vec == [])
assert(record_with_implicit_defaults.map == dict())
assert(record_with_implicit_defaults.some_bytes == b'')
assert(record_with_implicit_defaults.opt_int32 is None)
assert(record_with_implicit_defaults.custom_integer == 0)

assert(double_with_default() == 42)
assert(sum_with_default(1) == 1)
assert(sum_with_default(1, 2) == 3)

obj_with_defaults = ObjectWithDefaults()
assert(obj_with_defaults.add_to_num() == 42)
assert(obj_with_defaults.add_to_implicit_num() == 30)
assert(obj_with_defaults.add_to_implicit_num(1) == 31)

# Traits
class PyTestCallbackInterface(TestCallbackInterface):
    def do_nothing(self):
        pass

    def add(self, a, b):
        return a + b

    def optional(self, a):
        if a is None:
            return 0
        return a

    def with_bytes(self, rwb):
        return rwb.some_bytes

    def try_parse_int(self, value):
        if value == "force-unexpected-error":
            # raise an error that's not expected
            raise KeyError(value)
        try:
            return int(value)
        except BaseException:
            raise BasicError.InvalidInput()

    def callback_handler(self, h):
        v = h.take_error(BasicError.InvalidInput())
        return v

    def get_other_callback_interface(self):
        return PyTestCallbackInterface2()

class PyTestCallbackInterface2(OtherCallbackInterface):
    def multiply(self, a, b):
        return a * b

call_callback_interface(PyTestCallbackInterface())

# udl exposed functions with procmacro types.
assert get_one(None).inner == 0
assert get_bool(None) == MaybeBool.UNCERTAIN
assert get_object(None).is_heavy() == MaybeBool.UNCERTAIN
assert get_trait_with_foreign(None).name() == "RustTraitImpl"
assert get_externals(None).one is None

# values for enums without an explicit value are their index.
assert(MaybeBool.TRUE.value == 0)
assert(MaybeBool.FALSE.value == 1)
assert(MaybeBool.UNCERTAIN.value == 2)
# values with an explicit value should be that value.
assert(ReprU8.ONE.value == 1)
assert(ReprU8.THREE.value == 3)
#assert(ReprU8.FIVE.value == 5)

assert(get_mixed_enum(None) == MixedEnum.INT(1))
assert(get_mixed_enum(MixedEnum.NONE()) == MixedEnum.NONE())
assert(get_mixed_enum(MixedEnum.STRING("hello")) == MixedEnum.STRING("hello"))
assert(MixedEnum.STRING("hello")[0] == "hello")
assert(str(MixedEnum.STRING("hello")) == "MixedEnum.STRING('hello',)")

assert(MixedEnum.BOTH("hello", 1)[0] == "hello")
assert(MixedEnum.BOTH("hello", 1)[1] == 1)
assert(MixedEnum.BOTH("hello", 1)[:] == ('hello', 1))
assert(MixedEnum.BOTH("hello", 1)[-1] == 1)
assert(str(MixedEnum.BOTH("hello", 2)) == "MixedEnum.BOTH('hello', 2)")
assert(MixedEnum.STRING("").is_not_none())
assert(not MixedEnum.NONE().is_not_none())

# In #2270 we realized confusion about whether we generated our
# variant checker as, eg, `is_ALL()` vs `is_all()` so decided to do both.
assert(get_mixed_enum(MixedEnum.ALL("string", 2)).is_all())
assert(get_mixed_enum(MixedEnum.ALL("string", 2)).is_ALL())

assert(get_mixed_enum(MixedEnum.VEC(["string"])).is_VEC())
